<?xml version="1.0"?>
<launch>
  <!-- Использование симулированного времени -->
  <param name="use_sim_time" value="true"/>

  <!-- MAVROS для подключения к ArduPilot SITL -->
  <include file="$(find mavros)/launch/apm.launch">
    <arg name="fcu_url" value="udp://:14550@127.0.0.1:14557"/>
  </include>

  <!-- Загрузка URDF модели дрона -->
  <param name="robot_description" command="$(find xacro)/xacro '$(find basic_nav)/urdf/drone_urdf.urdf'"/>
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="state_publisher">
    <param name="publish_frequency" value="10.0"/>
  </node>

  <!-- Трансформации для лидара -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_lidar"
        args="0 0 0.2 0 0 0 base_link lidar_link 100"/>

  <!-- Публикация статической трансформации map -> odom -->
  <node pkg="basic_nav" type="map_to_odom.py" name="map_to_odom" output="screen"/>

  <!-- Загрузка карты -->
  <node pkg="map_server" type="map_server" name="map_server"
        args="$(find basic_nav)/maps/my_map.yaml"/>

  <!-- AMCL для локализации -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <param name="use_map_topic" value="true"/>
    <param name="base_frame_id" value="base_link"/>
    <param name="odom_frame_id" value="odom"/>
    <param name="global_frame_id" value="map"/>
    <rosparam file="$(find basic_nav)/config/amcl_params.yaml" command="load"/>
  </node>

  <!-- Настройка move_base для глобальной навигации -->
  <node pkg="move_base" type="move_base" name="move_base" output="screen">
    <rosparam file="$(find basic_nav)/config/costmap_params.yaml" command="load"/>
    <rosparam file="$(find basic_nav)/config/base_local_planner.yaml" command="load"/>
    <rosparam file="$(find basic_nav)/config/recovery_behaviors.yaml" command="load"/>
    <param name="planner_frequency" value="5.0"/>
    <param name="controller_frequency" value="10.0"/>
    <param name="base_global_planner" value="navfn/NavfnROS"/>
    <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS"/>
    <remap from="cmd_vel" to="/cmd_vel"/>
  </node>

  <!-- Запуск собственного глобального планировщика -->
  <node pkg="basic_nav" type="global_planner.py" name="drone_global_planner" output="screen">
    <param name="plan_frequency" value="1.0"/>
    <param name="altitude" value="1.5"/>
    <param name="use_markers" value="true"/>
  </node>

  <!-- Запуск локального планировщика с избеганием препятствий -->
  <node pkg="basic_nav" type="local_planner.py" name="drone_local_planner" output="screen">
    <rosparam file="$(find basic_nav)/config/local_planner_params.yaml" command="load"/>
  </node>

  <!-- Ретрансляция данных для локального планировщика -->
  <node pkg="topic_tools" type="relay" name="scan_relay"
        args="/drone1/scan /scan">
    <param name="queue_size" value="10"/>
    <param name="unreliable" value="true"/>
  </node>

  <!-- Преобразование данных пути в команды MAVROS -->
  <node pkg="basic_nav" type="path_to_setpoint.py" name="path_to_setpoint" output="screen"/>

  <!-- RViz для визуализации -->
  <node pkg="rviz" type="rviz" name="rviz"
        args="-d $(find basic_nav)/rviz/drone_navigation.rviz"/>
</launch> 